File: pydss/delivery/FuseObjImpl.py
from dss.control.impl.ControlElemImpl import (ControlElemImpl,)
from dss.common.impl.Utilities import (Utilities,)
from dss.delivery.Fuse import (Fuse,)
from dss.common.impl.DSSGlobals import (DSSGlobals,)
from dss.control.impl.ControlAction import (ControlAction,)
from dss.delivery.FuseObj import (FuseObj,)
from dss.delivery.impl.FuseImpl import (FuseImpl,)
# from org.apache.commons.math.complex.Complex import (Complex,)


class FuseObjImpl(ControlElemImpl, FuseObj):
    fuseCurve = None
    ratedCurrent = None
    delayTime = None
    monitoredElementName = None
    monitoredElementTerminal = None
    monitoredElement = None
    # Handle to control queue actions
    hAction = [None] * Fuse.FUSEMAXDIM
    # 0 = open 1 = close
    presentState = [None] * Fuse.FUSEMAXDIM
    readyToBlow = [None] * Fuse.FUSEMAXDIM
    # Offset for monitored terminal
    condOffset = None
    cBuffer = None

    def __init__(self, parClass, fuseName):
        super(FuseObjImpl, self)(parClass)
        self.setName(fuseName.toLowerCase())
        self.objType = parClass.getDSSClassType()
        self.setNPhases(3)
        # directly set conds and phases
        self.nConds = 3
        self.setNTerms(1)
        # this forces allocation of terminals and conductors in base class
        self.elementName = ''
        self.setControlledElement(None)
        self.elementTerminal = 1
        self.monitoredElementName = ''
        self.monitoredElementTerminal = 1
        self.monitoredElement = None
        self.fuseCurve = FuseImpl.getTccCurve('tlink')
        self.ratedCurrent = 1.0
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.Math.min(Fuse.FUSEMAXDIM, self.getNPhases())):
                break
            self.presentState[i] = ControlAction.CLOSE
        _1 = True
        i = 0
        while True:
            if _1 is True:
                _1 = False
            else:
                i += 1
            if not (i < self.Math.min(Fuse.FUSEMAXDIM, self.getNPhases())):
                break
            self.readyToBlow[i] = False
        _2 = True
        i = 0
        while True:
            if _2 is True:
                _2 = False
            else:
                i += 1
            if not (i < self.Math.min(Fuse.FUSEMAXDIM, self.getNPhases())):
                break
            self.hAction[i] = 0
        self.cBuffer = None
        # complex buffer
        self.objType = parClass.getDSSClassType()
        # CAP_CONTROL;
        self.initPropertyValues(0)
        # recalcElementData();

    def recalcElementData(self):
        devIndex = Utilities.getCktElementIndex(self.monitoredElementName)
        if devIndex >= 0:
            self.monitoredElement = DSSGlobals.activeCircuit.getCktElements().get(devIndex)
            self.setNPhases(self.monitoredElement.getNPhases())
            # force number of phases to be same
            if self.getNPhases() > Fuse.FUSEMAXDIM:
                DSSGlobals.doSimpleMsg('Warning: Fuse ' + self.getName() + ': Number of phases > max fuse dimension.', 404)
            if self.monitoredElementTerminal > self.monitoredElement.getNTerms():
                DSSGlobals.doErrorMsg('Fuse: \"' + self.getName() + '\"', 'Terminal no. \"' + '\" does not exist.', 'Re-specify terminal no.', 404)
            else:
                # sets name of i-th terminal's connected bus in fuse's bus list
                self.setBus(1, self.monitoredElement.getBus(self.monitoredElementTerminal))
                # TODO Check zero based indexing
                # allocate a buffer big enough to hold everything from the monitored element
                self.cBuffer = Utilities.resizeArray(self.cBuffer, self.monitoredElement.getYorder())
                self.condOffset = (self.monitoredElementTerminal - 1) * self.monitoredElement.getNConds()
                # for speedy sampling
        # Check for existence of controlled element
        devIndex = Utilities.getCktElementIndex(self.elementName)
        if devIndex >= 0:
            # both CktElement and monitored element must already exist
            self.setControlledBus(DSSGlobals.activeCircuit.getCktElements().get(devIndex))
            self.getControlledElement().setActiveTerminalIdx(self.elementTerminal)
            # make the 1st terminal active
            _0 = True
            i = 0
            while True:
                if _0 is True:
                    _0 = False
                else:
                    i += 1
                if (
                    not (i < self.Math.min(Fuse.FUSEMAXDIM, self.getControlledElement().getNPhases()))
                ):
                    break
                if self.getControlledElement().getConductorClosed(i):
                    # check state of i-th phase of active terminal
                    self.presentState[i] = ControlAction.CLOSE
                else:
                    self.presentState[i] = ControlAction.OPEN
            _1 = True
            i = 0
            while True:
                if _1 is True:
                    _1 = False
                else:
                    i += 1
                if not (i < self.getControlledElement().getNPhases()):
                    break
                self.hAction[i] = 0
            _2 = True
            i = 0
            while True:
                if _2 is True:
                    _2 = False
                else:
                    i += 1
                if (
                    not (i < self.Math.min(Fuse.FUSEMAXDIM, self.getControlledElement().getNPhases()))
                ):
                    break
                self.readyToBlow[i] = False
        else:
            self.setControlledElement(None)
            # element not found
            DSSGlobals.doErrorMsg('Fuse: \"' + self.getName() + '\"', 'CktElement Element \"' + self.elementName + '\" not found.', ' Element must be defined previously.', 405)

    def calcYPrim(self):
        """Always zero for a fuse."""
        # leave YPrims as null and they will be ignored
        pass

    def getCurrents(self, curr):
        """Get present value of terminal current."""
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.getNConds()):
                break
            curr[i] = Complex.ZERO

    def getInjCurrents(self, curr):
        """Returns injection currents."""
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.getNConds()):
                break
            curr[i] = Complex.ZERO

    def doPendingAction(self, phs, proxyHdl):
        """Do the action that is pending from last sample.

        Theoretically, there shouldn't be anything on the queue unless we have to do something.

        Only legal action is to open one phase.
        """
        if phs <= Fuse.FUSEMAXDIM:
            self.getControlledElement().setActiveTerminalIdx(self.elementTerminal)
            # set active terminal of CktElement to terminal 1
            _0 = self.presentState[phs]
            _1 = False
            while True:
                if _0 == self.CLOSE:
                    _1 = True
                    if self.readyToBlow[phs]:
                        # ignore if we became disarmed in meantime
                        self.getControlledElement().setConductorClosed(phs, False)
                        # open all phases of active terminal
                        Utilities.appendToEventLog('Fuse.' + self.getName(), 'Phase ' + String.valueOf.valueOf(phs) + ' Blown')
                        self.hAction[phs] = 0
                    break
                if True:
                    _1 = True
                    break
                break

    def interpretFuseAction(self, action):
        if self.getControlElement() is not None:
            self.getControlledElement().setActiveTerminalIdx(self.elementTerminal)
            # set active terminal
            _0 = action.toLowerCase()[0]
            _1 = False
            while True:
                if _0 == 'o':
                    _1 = True
                    self.getControlledElement().setConductorClosed(0, False)
                    # open all phases of active terminal   TODO Check zero based indexing
                    break
                if (_1 is True) or (_0 == 't'):
                    _1 = True
                    self.getControlledElement().setConductorClosed(0, False)
                    break
                if (_1 is True) or (_0 == 'c'):
                    _1 = True
                    self.getControlledElement().setConductorClosed(0, True)
                    # close all phases of active terminal
                    break
                break

    def sample(self):
        """Sample control quantities and set action times in control queue."""
        self.getControlledElement().setActiveTerminalIdx(self.elementTerminal)
        self.monitoredElement.getCurrents(self.cBuffer)
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if (
                not (i < self.Math.min(Fuse.FUSEMAXDIM, self.monitoredElement.getNPhases()))
            ):
                break
            if self.getControlledElement().getConductorClosed(i):
                # check state of phases of active terminal
                self.presentState[i] = ControlAction.CLOSE
            else:
                self.presentState[i] = ControlAction.OPEN
            if self.presentState[i] == ControlAction.CLOSE:
                tripTime = -1.0
                # Check phase trip, if any
                if self.fuseCurve is not None:
                    CMag = self.cBuffer[i].abs()
                    tripTime = self.fuseCurve.getTCCTime(CMag / self.ratedCurrent)
                if tripTime > 0.0:
                    if not self.readyToBlow[i]:
                        ckt = DSSGlobals.activeCircuit
                        # then arm for an open operation
                        self.hAction[i] = ckt.getControlQueue().push(ckt.getSolution().getIntHour(), ckt.getSolution().getDynaVars().t + tripTime + self.delayTime, i, 0, self)
                        self.readyToBlow[i] = True
                elif self.readyToBlow[i]:
                    # current has dropped below pickup and it hasn't blown yet
                    DSSGlobals.activeCircuit.getControlQueue().delete(self.hAction[i])
                    # delete the fuse blow action
                    self.readyToBlow[i] = False

    def dumpProperties(self, f, complete):
        super(FuseObjImpl, self).dumpProperties(f, complete)
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.getParentClass().getNumProperties()):
                break
            print '~ ' + self.getParentClass().getPropertyName()[i] + '=' + self.getPropertyValue(i)
        if complete:
            print 

    def getPropertyValue(self, index):
        return super(FuseObjImpl, self).getPropertyValue(index)

    def reset(self):
        """Reset to initial defined state."""
        if self.getControlledElement() is not None:
            _0 = True
            i = 0
            while True:
                if _0 is True:
                    _0 = False
                else:
                    i += 1
                if (
                    not (i < self.Math.min(Fuse.FUSEMAXDIM, self.getControlledElement().getNPhases()))
                ):
                    break
                self.presentState[i] = ControlAction.CLOSE
            _1 = True
            i = 0
            while True:
                if _1 is True:
                    _1 = False
                else:
                    i += 1
                if (
                    not (i < self.Math.min(Fuse.FUSEMAXDIM, self.getControlledElement().getNPhases()))
                ):
                    break
                self.readyToBlow[i] = False
            _2 = True
            i = 0
            while True:
                if _2 is True:
                    _2 = False
                else:
                    i += 1
                if (
                    not (i < self.Math.min(Fuse.FUSEMAXDIM, self.getControlledElement().getNPhases()))
                ):
                    break
                self.hAction[i] = 0
            self.getControlledElement().setActiveTerminalIdx(self.elementTerminal)
            # set active terminal
            self.getControlledElement().setConductorClosed(0, True)
            # close all phases of active terminal

    def initPropertyValues(self, arrayOffset):
        # FIXME: Private members in OpenDSS
        self.propertyValue[1] = ''
        # "element";
        self.propertyValue[2] = '1'
        # "terminal";
        self.propertyValue[3] = ''
        self.propertyValue[4] = '1'
        # "terminal";
        self.propertyValue[5] = 'Tlink'
        self.propertyValue[6] = '1.0'
        self.propertyValue[7] = '0'
        self.propertyValue[8] = ''
        super(FuseObjImpl, self).initPropertyValues(Fuse.NumPropsThisClass)

    def getFuseCurve(self):
        return self.fuseCurve

    def setFuseCurve(self, curve):
        self.fuseCurve = curve

    def getRatedCurrent(self):
        return self.ratedCurrent

    def setRatedCurrent(self, current):
        self.ratedCurrent = current

    def getDelayTime(self):
        return self.delayTime

    def setDelayTime(self, time):
        self.delayTime = time

    def getMonitoredElementName(self):
        return self.monitoredElementName

    def setMonitoredElementName(self, name):
        self.monitoredElementName = name

    def getMonitoredElementTerminal(self):
        return self.monitoredElementTerminal

    def setMonitoredElementTerminal(self, terminal):
        self.monitoredElementTerminal = terminal

    def getMonitoredElement(self):
        return self.monitoredElement

    def setMonitoredElement(self, element):
        self.monitoredElement = element

    def getHAction(self):
        return self.hAction

    def setHAction(self, action):
        self.hAction = action

    def getPresentState(self):
        return self.presentState

    def setPresentState(self, state):
        self.presentState = state

    def getReadyToBlow(self):
        return self.readyToBlow

    def setReadyToBlow(self, ready):
        self.readyToBlow = ready

    def getCondOffset(self):
        return self.condOffset

    def setCondOffset(self, offset):
        self.condOffset = offset

    def getCBuffer(self):
        return self.cBuffer

    def setCBuffer(self, buffer):
        self.cBuffer = buffer

File: core/dss-core/src/main/java/com/ncond/dss/delivery/impl/FuseObjImpl.java
package com.ncond.dss.delivery.impl;

import java.io.PrintStream;

import org.apache.commons.math.complex.Complex;

import com.ncond.dss.common.Bus;
import com.ncond.dss.common.Circuit;
import com.ncond.dss.common.impl.DSSCktElement;
import com.ncond.dss.common.impl.DSSClassImpl;
import com.ncond.dss.common.impl.DSSGlobals;
import com.ncond.dss.common.impl.Utilities;
import com.ncond.dss.control.impl.ControlAction;
import com.ncond.dss.control.impl.ControlElemImpl;
import com.ncond.dss.delivery.Fuse;
import com.ncond.dss.delivery.FuseObj;
import com.ncond.dss.general.TCC_CurveObj;

public class FuseObjImpl extends ControlElemImpl implements FuseObj {

	private TCC_CurveObj fuseCurve;

	private double ratedCurrent;

	private double delayTime;

	private String monitoredElementName;
	private int monitoredElementTerminal;
	private DSSCktElement monitoredElement;

	/* Handle to control queue actions */
	private int[] hAction = new int[Fuse.FUSEMAXDIM];
	/* 0 = open 1 = close */
	private ControlAction[] presentState = new ControlAction[Fuse.FUSEMAXDIM];
	private boolean[] readyToBlow = new boolean[Fuse.FUSEMAXDIM];

	/* Offset for monitored terminal */
	private int condOffset;
	private Complex[] cBuffer;

	public FuseObjImpl(DSSClassImpl parClass, String fuseName) {
		super(parClass);

		int i;

		setName(fuseName.toLowerCase());
		objType = parClass.getDSSClassType();

		setNPhases(3);  // directly set conds and phases
		nConds = 3;
		setNTerms(1);  // this forces allocation of terminals and conductors in base class

		elementName = "";
		setControlledElement(null);
		elementTerminal = 0;

		monitoredElementName = "";
		monitoredElementTerminal = 0;
		monitoredElement = null;

		fuseCurve = FuseImpl.getTccCurve("tlink");

		ratedCurrent = 1.0;

		for (i = 0; i < Math.min(Fuse.FUSEMAXDIM, getNPhases()); i++)
			presentState[i] = ControlAction.CLOSE;
		for (i = 0; i < Math.min(Fuse.FUSEMAXDIM, getNPhases()); i++)
			readyToBlow[i] = false;
		for (i = 0; i < Math.min(Fuse.FUSEMAXDIM, getNPhases()); i++)
			hAction[i] = 0;

		cBuffer = null;  // complex buffer

		objType = parClass.getDSSClassType(); //CAP_CONTROL;

		initPropertyValues(0);

		//recalcElementData();
	}

	@Override
	public void recalcElementData() {
		int i;

		int devIndex = Utilities.getCktElementIndex(monitoredElementName);
		if (devIndex >= 0) {
			monitoredElement = (DSSCktElement) DSSGlobals.activeCircuit.getCktElements().get(devIndex);
			setNPhases( monitoredElement.getNPhases() );  // force number of phases to be same
			if (getNPhases() > Fuse.FUSEMAXDIM)
				DSSGlobals.doSimpleMsg("Warning: Fuse "+getName()+": Number of phases > max fuse dimension.", 404);
			if (monitoredElementTerminal >= monitoredElement.getNTerms()) {
				DSSGlobals.doErrorMsg("Fuse: \"" + getName() + "\"",
						"Terminal no. \"" +"\" does not exist.",
						"Re-specify terminal no.", 404);
			} else {
				// sets name of i-th terminal's connected bus in fuse's bus list
				setBus(0, monitoredElement.getBus(monitoredElementTerminal));
				// allocate a buffer big enough to hold everything from the monitored element
				cBuffer = Utilities.resizeArray(cBuffer, monitoredElement.getYorder());
				condOffset = monitoredElementTerminal * monitoredElement.getNConds();  // for speedy sampling
			}
		}

		/* Check for existence of controlled element */

		devIndex = Utilities.getCktElementIndex(elementName);
		if (devIndex >= 0) {  // both CktElement and monitored element must already exist
			setControlledBus( (Bus) DSSGlobals.activeCircuit.getCktElements().get(devIndex) );
			getControlledElement().setActiveTerminalIdx(elementTerminal);  // make the 1st terminal active

			for (i = 0; i < Math.min(Fuse.FUSEMAXDIM, getControlledElement().getNPhases()); i++) {
				if (getControlledElement().getConductorClosed(i)) {  // check state of i-th phase of active terminal
					presentState[i] = ControlAction.CLOSE;
				} else {
					presentState[i] = ControlAction.OPEN;
				}
			}
			for (i = 0; i < getControlledElement().getNPhases(); i++)
				hAction[i] = 0;
			for (i = 0; i < Math.min(Fuse.FUSEMAXDIM, getControlledElement().getNPhases()); i++)
				readyToBlow[i] = false;
		} else {
			setControlledElement(null);  // element not found
			DSSGlobals.doErrorMsg("Fuse: \"" + getName() + "\"", "CktElement Element \""+ elementName + "\" not found.",
					" Element must be defined previously.", 405);
		}
	}

	/**
	 * Always zero for a fuse.
	 */
	@Override
	public void calcYPrim() {
		// leave YPrims as null and they will be ignored
	}

	/**
	 * Get present value of terminal current.
	 */
	@Override
	public void getCurrents(Complex[] curr) {
		for (int i = 0; i < getNConds(); i++)
			curr[i] = Complex.ZERO;
	}

	/**
	 * Returns injection currents.
	 */
	@Override
	public void getInjCurrents(Complex[] curr) {
		for (int i = 0; i < getNConds(); i++)
			curr[i] = Complex.ZERO;
	}

	/**
	 * Do the action that is pending from last sample.
	 *
	 * Theoretically, there shouldn't be anything on the queue unless we have to do something.
	 *
	 * Only legal action is to open one phase.
	 */
	@Override
	public void doPendingAction(int phs, int proxyHdl) {
		if (phs <= Fuse.FUSEMAXDIM) {
			getControlledElement().setActiveTerminalIdx(elementTerminal);  // set active terminal of CktElement to terminal 1
			switch (presentState[phs]) {
			case CLOSE:
				if (readyToBlow[phs]) {  // ignore if we became disarmed in meantime
					getControlledElement().setConductorClosed(phs, false);  // open all phases of active terminal
					Utilities.appendToEventLog("Fuse." + getName(), "Phase "+String.valueOf(phs)+" Blown");
					hAction[phs] = 0;
				}
				break;
			default:
				// do nothing
				break;
			}
		}
	}

	public void interpretFuseAction(String action) {
		if (getControlElement() != null) {
			getControlledElement().setActiveTerminalIdx(elementTerminal);  // set active terminal
			switch (action.toLowerCase().charAt(0)) {
			case 'o':
				getControlledElement().setConductorClosed(0, false);  // open all phases of active terminal   TODO Check zero based indexing
				break;
			case 't':
				getControlledElement().setConductorClosed(0, false);
				break;
			case 'c':
				getControlledElement().setConductorClosed(0, true);  // close all phases of active terminal
				break;
			}
		}
	}

	/**
	 * Sample control quantities and set action times in control queue.
	 */
	@Override
	public void sample() {
		double CMag;
		double tripTime;

		getControlledElement().setActiveTerminalIdx(elementTerminal);
		monitoredElement.getCurrents(cBuffer);

		for (int i = 0; i < Math.min(Fuse.FUSEMAXDIM, monitoredElement.getNPhases()); i++) {
			if (getControlledElement().getConductorClosed(i)) {  // check state of phases of active terminal
				presentState[i] = ControlAction.CLOSE;
			} else {
				presentState[i] = ControlAction.OPEN;
			}

			if (presentState[i] == ControlAction.CLOSE) {
				tripTime = -1.0;

				/* Check phase trip, if any */

				if (fuseCurve != null) {
					CMag = cBuffer[i].abs();
					tripTime = fuseCurve.getTCCTime(CMag / ratedCurrent);
				}

				if (tripTime > 0.0) {
					if (!readyToBlow[i]) {
						Circuit ckt = DSSGlobals.activeCircuit;
						// then arm for an open operation
						hAction[i] = ckt.getControlQueue().push(ckt.getSolution().getIntHour(),
								ckt.getSolution().getDynaVars().t + tripTime + delayTime, i, 0, this);
						readyToBlow[i] = true;
					}
				} else {
					if (readyToBlow[i]) {
						// current has dropped below pickup and it hasn't blown yet
						DSSGlobals.activeCircuit.getControlQueue().delete(hAction[i]);  // delete the fuse blow action
						readyToBlow[i] = false;
					}
				}
			}
		}
	}

	@Override
	public void dumpProperties(PrintStream f, boolean complete) {
		super.dumpProperties(f, complete);

		for (int i = 0; i < getParentClass().getNumProperties(); i++)
			f.println("~ " + getParentClass().getPropertyName()[i] + "=" + getPropertyValue(i));

		if (complete)
			f.println();
	}

	@Override
	public String getPropertyValue(int index) {
		return super.getPropertyValue(index);
	}

	/**
	 * Reset to initial defined state.
	 */
	@Override
	public void reset() {
		int i;

		if (getControlledElement() != null) {
			for (i = 0; i < Math.min(Fuse.FUSEMAXDIM, getControlledElement().getNPhases()); i++)
				presentState[i] = ControlAction.CLOSE;
			for (i = 0; i < Math.min(Fuse.FUSEMAXDIM, getControlledElement().getNPhases()); i++)
				readyToBlow[i] = false;
			for (i = 0; i < Math.min(Fuse.FUSEMAXDIM, getControlledElement().getNPhases()); i++)
				hAction[i] = 0;
			getControlledElement().setActiveTerminalIdx(elementTerminal);  // set active terminal
			getControlledElement().setConductorClosed(0, true);            // close all phases of active terminal
		}
	}

	@Override
	public void initPropertyValues(int arrayOffset) {
		setPropertyValue(0, "");   // "element";
		setPropertyValue(1, "1");  // "terminal";
		setPropertyValue(2, "");
		setPropertyValue(3, "1");  // "terminal";
		setPropertyValue(4, "Tlink");
		setPropertyValue(5, "1.0");
		setPropertyValue(6, "0");
		setPropertyValue(7, "");

		super.initPropertyValues(Fuse.NumPropsThisClass - 1);
	}

	// FIXME: Private members in OpenDSS

	public TCC_CurveObj getFuseCurve() {
		return fuseCurve;
	}

	public void setFuseCurve(TCC_CurveObj curve) {
		fuseCurve = curve;
	}

	public double getRatedCurrent() {
		return ratedCurrent;
	}

	public void setRatedCurrent(double current) {
		ratedCurrent = current;
	}

	public double getDelayTime() {
		return delayTime;
	}

	public void setDelayTime(double time) {
		delayTime = time;
	}

	public String getMonitoredElementName() {
		return monitoredElementName;
	}

	public void setMonitoredElementName(String name) {
		monitoredElementName = name;
	}

	public int getMonitoredElementTerminal() {
		return monitoredElementTerminal;
	}

	public void setMonitoredElementTerminal(int terminal) {
		monitoredElementTerminal = terminal;
	}

	public DSSCktElement getMonitoredElement() {
		return monitoredElement;
	}

	public void setMonitoredElement(DSSCktElement element) {
		monitoredElement = element;
	}

	public int[] getHAction() {
		return hAction;
	}

	public void setHAction(int[] action) {
		this.hAction = action;
	}

	public ControlAction[] getPresentState() {
		return presentState;
	}

	public void setPresentState(ControlAction[] state) {
		presentState = state;
	}

	public boolean[] getReadyToBlow() {
		return readyToBlow;
	}

	public void setReadyToBlow(boolean[] ready) {
		readyToBlow = ready;
	}

	public int getCondOffset() {
		return condOffset;
	}

	public void setCondOffset(int offset) {
		condOffset = offset;
	}

	public Complex[] getCBuffer() {
		return cBuffer;
	}

	public void setCBuffer(Complex[] buffer) {
		cBuffer = buffer;
	}

}
